### PR TITLE
Add GREP: Coding guidelines

### PR DESCRIPTION
**Pull request by mbr0wn**

### CODE REVIEW

File: grep-0001-coding-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#include "module_specific.hpp"
+#include <gnuradio/block.h>
+#include <fftw3.h>
+#include <boost/shared_ptr.hpp>
+#include <mutex>
+```
+
+* Feel free to use modern C/C++ features even if they were not used before.
+  Make sure they work with the compilers and dependencies which are set for the
+  version of UHD the commit will be made upon. The GNU Radio CI system will be
```
Comment (n-west) [inline grep-0001-coding-guidelines.md]: s/UHD/GNU Radio/

**Comment by giladbeeri** [comment]
Should comments be written here, or in the diff/PR view (as source-code comments)?

**Comment by mbr0wn** [comment]
@giladbeeri Both is fine, I'd recommend complaints/comments on specific paragraphs on the source-code view, and general comments here.

Once this is merged, comments should probably best be submitted as issues. Thanks for participating!

File: grep-0001-coding-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# GREP [XXXX] -- [Title]
```
Comment (giladbeeri) [inline grep-0001-coding-guidelines.md]: not to be picky, but set an example by filling the template :)

File: grep-0001-coding-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+To summarize, consistency is extremely useful, and coding guidelines help ensure
+consistency. We assume that all contributors are mature enough to at least try
+and follow them.
+
+## Description
+
+### General Coding Guidelines
+
+* Never submit code with trailing whitespace.
```
Comment (giladbeeri) [inline grep-0001-coding-guidelines.md]: I'll add here: "always end a file with a newline" for the same terminal-usage reason you used below

File: grep-0001-coding-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+* Avoid Boost where possible.
+* Don't use Boost's sleep functions. Prefer `std::chrono` functions.
+
+## Python-specific Guidelines
+
+* Follow the suggestions in PEP8 (https://www.python.org/dev/peps/pep-0008/)
+  and PEP257 (https://www.python.org/dev/peps/pep-0257/). The former is about
+  code layout in general, the latter about docstrings.
+* Keep Python code compatible with Py2k and Py3k, where possible.
```
Comment (giladbeeri) [inline grep-0001-coding-guidelines.md]: I assume it's better to be explicit and set a standard for the supported Python version, which in 2.X should be 2.7 IMO.

File: grep-0001-coding-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+* Pylint is good tool for helping with following code guidelines. It's very
+  fussy though, so don't get too worked up about following its suggestions.
+
+## Revision Control Hygiene
+
+* In this repository, we almost always use fast-forward merges, and no merge
```
Comment (giladbeeri) [inline grep-0001-coding-guidelines.md]: In my experience, when writing guidelines, it's better to give the reasoning behind, so readers know "ah, that's why they insist on that". See your terminal-usage example above, which is great.

What is the reasoning here (no merge commits)?

File: grep-0001-coding-guidelines.md (hunk 2 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: Whoops.

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: So, this is one of the guidelines I just copied over from UHD. There's two things here:
- Fast-forward merges: These are best to keep development linear, and add new features always on top. Also, the best way to avoid merge conflicts (but that's already somewhat subjective). This one I feel pretty strongly about.
- No merge commits: When doing FF merges, they are simply not necessary and clutter the history. However, they can also be nice because they give some indication of the source branch. git default is to not do merge commits on ff merges, but we could do this differently.

Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: ...but excellent point about adding rationales. I think we could use more of those.

File: grep-0001-coding-guidelines.md (hunk 4 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: That's a good point; we actually need to lock down Python versions actually. For the upcoming 3.8 release, we might just drop Py2k support altogether and then this point needs some rewriting anyway.

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (giladbeeri) [reply grep-0001-coding-guidelines.md]: Regarding the 2nd issue, I agree with 2 of your statements: if doing FF merges, merge commits are less useful, but still nice. But, it is completely based on the 1st issue.

I understand the history cleanliness that comes with FF merges, but for them to work, it means the collaborator needs to rebase his branch (assuming the source branch, e.g., master, progressed since branching), which usually causes more conflicts, and forces her/him to modify the history - an unpleasant action of several collaborators work on the same branch, and in general, and in addition, you lose information.

The approach I find most useful:

* Always have a merge commit.
* Don't enforce rebasing. If a fast-forward is possible (master hasn't evolved), so be it. Otherwise, fast-forward isn't a must.
* Collaborator solves conflicts by merging master into her own branch when they finish (actually, preferred on an ongoing basis, as it makes solving conflicts much easier) and solve all of the conflicts. That way, when the project maintainer merges back into master, there are no conflicts.
* Commit comment guidelines (prefixes, components, etc.) aren't strictly enforced on commits that belong to a branch, but are encouraged for important commits. This allows the developer more freedom and flexibility, e.g., commit something because they are going to continue next week and don't want to risk losing it, so they just use a message of "temp", "comment" if it's a slight comment change, "docs", etc.
* Commit comment guidelines are enforced on merge commits on master.
* When developers want to see the history with its most important commits (AKA only the merge commits), they use `git log --first-parent`. They get nice, linear, well-documented history. When they want to see the small details and try to simulate the way of thinking of the developer, they can check the branch commits.

Less related to this issue:

* Merge commits should include the branch name and a link to GitHub's PR.
* When solving conflicts and committing, it's best to document in the commit message the files that required conflict resolution (I think you get it as a commented-out list anyway) - this help in tracking bugs due to bad conflict resolutions ("oh, he merged and solved a conflict, but didn't realize line A should come before B").

File: grep-0001-coding-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
+# GREP 0001 -- Coding Guidelines
+
+- Original Author: Martin Braun <martin@gnuradio.org>,
+                   Marcus Müller <marcus@hostalia.de>
```
Comment (marcusmueller) [inline grep-0001-coding-guidelines.md]: can you put my gnuradio.org address here?

File: grep-0001-coding-guidelines.md (hunk 4 above)
Comment (marcusmueller) [reply grep-0001-coding-guidelines.md]: Yes, I'd argue that this document would apply to 3.8 and further on, anyway. Py2k statements might optimally be found under "if you're changing lecacy (pre-3.8) code…", or so?

File: grep-0001-coding-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+#include <gnuradio/block.h>
+#include <fftw3.h>
+#include <boost/shared_ptr.hpp>
+#include <mutex>
+```
+
+* Feel free to use modern C/C++ features even if they were not used before.
+  Make sure they work with the compilers and dependencies which are set for the
+  version of GNU Radio the commit will be made upon. The GNU Radio CI system
+  will be able to confirm this. (Note: C++11 features are available starting
```
Comment (marcusmueller) [inline grep-0001-coding-guidelines.md:95]: There should be a definite statement here: GNU Radio 3.7 uses C++98, from 3.8 on we use C++11.

Especially the stanza about preferring C++ constructs over boost could use that as a backing argument.

**Comment by marcusmueller** [comment]
Do we want to talk about code *style* / mechanics in this document? This draft touches onto this only with 

* line endings and
* include order;

but that's feels like it's two relatively arbitrary picks¹.

I'd argue that this either needs to be a style guide, defining things like indents, naming, or these two feel a bit out of place.

I use the [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html) as inspiration (obviously, we're not LLVM and don't agree on "don't use `dynamic_cast`" nor "don't use exceptions", but aside from that, people that write compilers and their runtimes might not be the worst reference). Also, LLVM is decidedly non-GNU style and I find it relatively pleasing to read.

The strength of the LLVM guidelines is that they're C++11, something that I, as someone who's spending most of his time in pre-C++11 code, still need to come up with good practices for.

Also, it comes with exactly the indentation rules I like.

<hr>
¹with the include order actually having semantic effects on things with lesser sanity, but: for example, fftw explicitly says it behaves differently when system `complex.h` was included before `fftwf.h`, so meh, I actually find this nice to read, but not how I want my system to actually process the files. I'm actually not a big fan of that rule, as `#include` order has a semantic effect, just like writing C++, and you don't write bad programs just because some dude said that all variables used should be named in a certain order. I'm still for having this rule in there, but it's really a "low-prio, know what you're doing" rule.

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: I'd like to hear more people's opinions on this. Like I said, this was a base for discussion. A couple of things on your comments though:

- Your approach still doesn't aid bisecting
- When commits are submitted for merge, the commit message should be enforced (because they end up in the final history). People's intermediate commits need to be cleaned up before submitting a PR. And before submitting your PR, you should rebase anyway, that's just good manners. The developer doesn't necessarily have to do the final rebase, the maintainer can still do that (it's even built into github UI these days) and resolving merge conflicts is either easy (in which case the maintainer can do it) or not (in which case the original developer needs to do it anyway, regardless of it being a rebase or a regular merge).

Basically, you're saying all the hard work needs to be with the maintainer, and the developer can do pretty much what they want, including formatting commit messages -- that's not acceptable.

Comment (giladbeeri) [reply grep-0001-coding-guidelines.md]: Why does my approach require rebasing? I might be missing something.

**Comment by mbr0wn** [comment]
@marcusmueller Yes, please up more stuff. As long as this document doesn't become overburdening on either maintainer or contributor, we can talk about any changes! And yes, the include order and line endings are arbitrary picks.

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: Your approach does not require rebasing, but that's what I'm complaining about 😄 Before a PR gets submitted, people should rebase, that's just good manners and is the best way to ensure no merge conflicts. Criss-cross merges (i.e., first merging master into your branch) means we end up with even more merge commits, and bisects become even harder.

I like some of your points, though. If we have merge commits, then references to the source branch are a good idea. Maybe this would be better served on the mailing list, where more people see this discussion and can chime in.

File: grep-0001-coding-guidelines.md (hunk 7 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md:95]: We can have different guidelines per branch, too, with different versions. One problem that we might have is that we don't necessarily have C++11, but we have all features that are supported by the common set of C++ features between all compilers that we officially support. This is particularly true for 3.7, which actually uses a superset of C++98, but not exactly C++11.

**Comment by mbr0wn** [comment]
If we don't want all LLVM guidelines, but some, do we copy them out, or reference LLVM and say "use good judgement"?

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (marcusmueller) [reply grep-0001-coding-guidelines.md]: Regarding rebases:

I was very tempted to phrase it quite a bit stricter than what @mbr0wn put into that document. Again, this could do very well with a rationale, put into a separate bullet point:

> * Before opening a request for merging your code, make sure to rebase your changes against the current state of the branch you forked off your own branch (typically, that's `master`, so that procedure would amount to `git fetch origin; git rebase origin/master`). <br> That way, you keep control of how your code integrates into the common code base! If everything goes well, rebasing requires absolutely no interaction on your side. If something goes wrong, fix it like you'd fix a merge conflict. It's the same effort as fixing a merge conflict, and avoids one of those later on, but puts you in command over the order of lines, and enables you (who's the most knowledgeable expert on your own code) to spot problems that through naive merging by a maintainer would have become a bug.

**Comment by mbr0wn** [comment]
@giladbeeri When we merge this, it'll still be in "Draft" status and I would love to see you continuing the discussion.

File: grep-0001-coding-guidelines.md (hunk 5 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md]: I've added this paragraph in the latest version.

**Comment by marcusmueller** [comment]
Re: Style: This doesn't question the GREP in its current state, and is probably a matter complicated enough to amend the GREP through a separate PR. Will do that once we put this GREP into the draft state.

File: grep-0001-coding-guidelines.md (hunk 7 above)
Comment (awalls-cx18) [reply grep-0001-coding-guidelines.md:95]: Just because C++11 supports something, doesn't mean MSVC gets it right (and, or, not, vs. &&, ||, !): https://github.com/gnuradio/gnuradio/pull/1571

Comment (mbr0wn) [reply grep-0001-coding-guidelines.md:95]: Exactly! Hence my reference to the CI being the golden reference...  except we don't actually test MSVC.

File: grep-0001-coding-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+and follow them.
+
+## Description
+
+### General Coding Guidelines
+
+* Never submit code with trailing whitespace.
+* Code is read more often than it's written. Code readability is thus something
+  worth optimizing for.
+* Try and keep line lengths to 79 characters, unless readability suffers. We
```
Comment (smunaut) [inline grep-0001-coding-guidelines.md]: Really ?  79 ?
I mean it's 2018 and I can have side by side terminal with 140 char each ... now I agree my case might be extreme, but wouldn't taking a more modern value like 100 char width be more future looking ? Especially for C++ where lines do tend to get a bit long ?

Comment (marcusmueller) [reply grep-0001-coding-guidelines.md]: yeah, 79 is definitely at the lower end of useful limits (makes a lot of sense for other things, but not especially C++ in my experience). With a two- to four spaces indention depth, I'd argue I usually can live with up to ca 110 characters per line without losing the feeling that things on that line should be on that line. But that's definitely the uppermost boundary for length of lines where I'd be willing to not complain as a maintainer – so, 100 sounds fine to me.

**Comment by noc0lour** [comment]
Was wondering about the missing coding style part as well. Conform with existing style only works if there is already existing style. For new files we need to pinpoint how we want them to look like.

**Comment by mbr0wn** [comment]
@marcusmueller @noc0lour Do you want to start writing that?

**Comment by marcusmueller** [comment]
@mbr0wn as a separate GREP?

**Comment by mbr0wn** [comment]
No, as part of this one.

On 18 Feb 2018 08:36, "Marcus Müller" <notifications@github.com> wrote:

@mbr0wn <https://github.com/mbr0wn> as a separate Grep?

**Comment by giladbeeri** [comment]
If a language has a widely-accepted and used coding conventions (e.g., PEP8), I think it's better to "inherit" those by referral and modify/extend where needed instead of copying 99%.

If a language hasn't one (e.g., C++), maybe the discussion can be around several popular ones, aiming to choose one as a basis (LLVM, Google, ...).

File: grep-0001-coding-guidelines.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  will be able to confirm this. (Note: C++11 features are available starting
+  with GNU Radio version 3.8).
+
+
+## Boost-specific Guidelines
+
+* Avoid Boost where possible (see C++ guidelines for rationale).
+* Don't use Boost's sleep functions. Prefer `std::chrono` functions.
+
+## Python-specific Guidelines
```
Comment (jdemel) [inline grep-0001-coding-guidelines.md:104]: Since there is a Boost specific section, should we also have a NumPy/SciPy related section? At least this is something that pops up regularly on the bugtracker.
Something like:
* Try to avoid SciPy and make use of NumPy constructs.
* SciPy is optional, your code should have a non-SciPy fallback. Such guidelines would've helped me back in the day.

**Comment by noc0lour** [comment]
@giladbeeri in the Python section this GREP is explicitly referring to PEP8. For a C++ section @marcusmueller already suggested lending heavily from LLVM. We just need to nail down the parts we want to diverge from LLVM.

File: grep-0001-coding-guidelines.md (hunk 9 above)
Comment (mbr0wn) [reply grep-0001-coding-guidelines.md:104]: That's a great idea. Would you volunteer to write something up?

**Comment by mbr0wn** [comment]
@noc0lour: @giladbeeri was referring (I think) to my question *how* exactly we do that, and it's a good way forward.

Keep in my mind that we already refer to the core guidelines in the current draft. If we add others as reference, we will most likely run into conflicts.

**Comment by mbr0wn** [comment]
@giladbeeri @marcusmueller @noc0lour @smunaut @dkozel Moving the open items to issues.
